web/src/BulkApiButton.tsx

```diff
@@ -0,0 +1,363 @@
+
+  &:hover,
+  &:active,
+  &:focus {
+    background-color: ${Color.grayLighter};
+    color: ${Color.blue};
+  }
+
+  &.Mui-disabled {
+    border-color: ${Color.grayLight};
+    color: ${Color.gray6};
+  }
+
+  /* Use shared styles with ApiButton */
+  ${confirmingButtonStateMixin}
+  ${confirmingButtonGroupBorderMixin}
+`
+
+const BulkButtonGroup = styled(ButtonGroup)<{ disabled?: boolean }>`
+  ${(props) =>
+    props.disabled &&
+    `
+    cursor: not-allowed;
+  `}
+
+  /* Adjust the border radius of buttons to achieve the "rounded bar" look */
+  &.firstButtonGroupInRow {
+    ${BulkButtonElementRoot} {
+      border-top-right-radius: 0;
+      border-bottom-right-radius: 0;
+      border-right: 0;
+    }
+  }
+
+  &.middleButtonGroupInRow {
+    ${BulkButtonElementRoot} {
+      border-radius: 0;
+      border-right: 0;
+    }
+  }
+
+  &.lastButtonGroupInRow {
+    ${BulkButtonElementRoot} {
+      border-top-left-radius: 0;
+      border-bottom-left-radius: 0;
+    }
+  }
+`
+
+// Helpers
+export function canButtonBeToggled(
+  uiButton: UIButton,
+  targetToggleState?: ApiButtonToggleState
+) {
+  const toggleInput = uiButton.spec?.inputs?.find(
+    (input) => input.name === UIBUTTON_TOGGLE_INPUT_NAME
+  )
+
+  if (!toggleInput) {
+    return false
+  }
+
+  if (!targetToggleState) {
+    return true
+  }
+
+  const toggleValue = toggleInput.hidden?.value
+
+  // A button can be toggled if it's state doesn't match the target state
+  return toggleValue !== undefined && toggleValue !== targetToggleState
+}
+
+/**
+ * A bulk button can be toggled if some UIButtons have values that don't
+ * match the target toggle state.
+ * ex: some buttons are off and target toggle is on => bulk button can be toggled
+ * ex: all buttons are on and target toggle is on   => bulk button cannot be toggled
+ * ex: all buttons are not toggle buttons           => bulk button cannot be toggled
+ */
+export function canBulkButtonBeToggled(
+  uiButtons: UIButton[],
+  targetToggleState: ApiButtonToggleState
+) {
+  const individualButtonsCanBeToggled = uiButtons.map((b) =>
+    canButtonBeToggled(b, targetToggleState)
+  )
+
+  return individualButtonsCanBeToggled.some(
+    (canBeToggled) => canBeToggled === true
+  )
+}
+
+function isBulkButtonDisabled(
+  uiButtons: UIButton[],
+  targetToggleState?: ApiButtonToggleState
+) {
+  // Bulk button is disabled if there are no UIButtons to trigger
+  if (uiButtons.length === 0) {
+    return true
+  }
+
+  // If there's a target toggle state, calculate whether the bulk button
+  // is disabled based on the toggle values of all UIButtons
+  if (targetToggleState) {
+    const isDisabled = !canBulkButtonBeToggled(uiButtons, targetToggleState)
+    return isDisabled
+  }
+
+  return false
+}
+
+async function bulkUpdateButtonStatus(uiButtons: UIButton[]) {
+  try {
+    await Promise.all(uiButtons.map((button) => updateButtonStatus(button, {})))
+  } catch (err) {
+    throw err
```

Comment:
Any reason for the try-catch if we're re-throwing as-is?

Reply:
that's a great question! my native promise knowledge is a little rusty, but i thought that it's best practice to try/catch here, even though i'm expecting the parent component to do the error handling. (if nothing else, it does explicitly document that expectation. 🤷🏻‍♂️ happy to be corrected if i'm wrong!)